context_interp/model.py

```diff
@@ -0,0 +1,104 @@
+from common.models import RestorableNetwork
+from common.forward_warp.forward_warp import forward_warp
+
+
+class ContextInterp:
+    def __init__(self, name='context_interp'):
+        """
+        :param name: Str. For Tf variable scoping.
+        """
+        self.name = name
+        self.gridnet = GridNet([32, 64, 96], 6, num_output_channels=3)
+        self.laplacian_pyramid = LaplacianPyramid(5)
+        self.pwcnet = PWCNet()
+        self.feature_extractor = Vgg19Features()
+        self.feature_extractor.load_pretrained_weights()
+
+    def get_forward(self, image_a, image_b, t, reuse_variables=tf.AUTO_REUSE):
+        """
+        :param image_a: Tensor of shape [batch_size, H, W, 3].
+        :param image_b: Tensor of shape [batch_size, H, W, 3].
+        :param t: Float. Specifies the interpolation point (i.e 0 for image_a, 1 for image_b).
+        :return: interpolated: The interpolated image. Tensor of shape [batch_size, H, W, 3].
+                 warped_a_b: Image and features from a forward-flowed towards b, before synthesis.
+                             The first 3 channels are the image.
+                 warped_b_a: Image and features from b forward-flowed towards a, before synthesis.
+                             The first 3 channels are the image.
+                 flow_a_b: Flow from a to b (centered at a).
+                 flow_b_a: Flow from b to a (centered at b).
+        """
+        with tf.variable_scope(self.name, reuse=reuse_variables):
+            image_a_contexts = self.feature_extractor.get_context_features(image_a)
+            image_b_contexts = self.feature_extractor.get_context_features(image_b)
+
+            # TODO Add instance normalization. Described in 3.3 of https://arxiv.org/pdf/1803.10967.pdf.
+
+            # Get a->b and b->a flows from PWCNet.
+            flow_a_b, _ = self.pwcnet.get_forward(image_a, image_b)
+            flow_b_a, _ = self.pwcnet.get_forward(image_b, image_a)
+
+            features_a = tf.concat([image_a, image_a_contexts], axis=-1)
+            features_b = tf.concat([image_b, image_b_contexts], axis=-1)
+
+            # Warp images and their contexts from a->b and from b->a.
+            warped_a_b = forward_warp(features_a, t * flow_a_b)
+            warped_b_a = forward_warp(features_b, (1.0 - t) * flow_b_a)
+
+            # Feed into GridNet for final synthesis.
+            warped_combined = tf.concat([warped_a_b, warped_b_a], axis=-1)
+            warped_combined = tf.stop_gradient(warped_combined)
+            synthesized, _, _, _ = self.gridnet.get_forward(warped_combined, training=True)
+            return synthesized, warped_a_b, warped_b_a, flow_a_b, flow_b_a
+
+    def load_pwcnet_weights(self, pwcnet_weights_path, sess):
+        """
+        Loads pre-trained PWCNet weights.
+        For this to work:
+            - It must be called after get_forward.
+            - get_forward must not have been called in an enclosing variable scope.
```

Comment:
It's still true, since scope_prefix only uses self.name. An enclosing scope during variable definitions would not work well with this.

Reply:
The fact that scope_prefix only uses self.name is a self-fulfilling prophecy -- that's just how you wrote this function. Wouldn't it be more useful to add a comment on how changes to the topography affect this function?

Reply:
I wrote a function and stated the pre-reqs for using it, I don't see anything wrong with that in itself. I have updated this function so that it doesn't have that particular requirement though.